### PR TITLE
Cryptology Lab1

### PR DESCRIPTION
**Pull request by Pilipets**
1. Implemented the Fermat primality test.
2. Implemented the Miller–Rabin primality test
3. Implemented the Karatsuba Multiplication algorithm.
4. Implemented the Extended Euclidean algorithm.
5. Implemented Montgomery modular multiplication, modular exponentiation.
6. Implemented fast Modular exponentiation.
7. Added tests for each algorithm.

All numbers are BigIntegers with bit length >= 512.

### CODE REVIEW

File: part2/Crypotology(Lab1)/src/main/java/lab1/ExtendedEuclidean.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+public class ExtendedEuclidean {
+    // Assumes a and b are positive
+    public static BigInteger[] xgcd(BigInteger a, BigInteger b) {
+        BigInteger[] qrem;
+        BigInteger[] result = new BigInteger[3];
+        BigInteger x = a, y=b;
+        BigInteger x0 = BigInteger.ONE, x1 = BigInteger.ZERO;
+        BigInteger y0 = BigInteger.ZERO, y1 = BigInteger.ONE;
+        while (true){
+            qrem = x.divideAndRemainder(y); x = qrem[1];
```
Comment (kvaka313) [inline part2/Crypotology(Lab1)/src/main/java/lab1/ExtendedEuclidean.java]: Please do not ignore Java Code Convention

File: part2/Crypotology(Lab1)/src/main/java/lab1/FermatTest.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
+package lab1;
+
+import java.math.BigInteger;
+import java.util.Random;
+
+// https://en.wikipedia.org/wiki/Fermat_primality_test
+public class FermatTest
+{
```
Comment (kvaka313) [inline part2/Crypotology(Lab1)/src/main/java/lab1/FermatTest.java]: Please use jUnit for tests

File: part2/Crypotology(Lab1)/src/main/java/lab1/Karatsuba.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+        BigInteger ac    = karatsubaMultiply(a, c);
+        BigInteger bd    = karatsubaMultiply(b, d);
+        BigInteger abcd  = karatsubaMultiply(a.add(b), c.add(d));
+
+        return ac.add(abcd.subtract(ac).subtract(bd).shiftLeft(N)).add(bd.shiftLeft(2*N));
+    }
+
+
+    public static void main(String[] args) {
```
Comment (kvaka313) [inline part2/Crypotology(Lab1)/src/main/java/lab1/Karatsuba.java]: exclude this method

File: part2/Crypotology(Lab1)/src/main/java/lab1/ExtendedEuclidean.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            x1 = x1.subtract(y1.multiply(qrem[0]));
+            if (x.equals(BigInteger.ZERO)) {result[0]=y; result[1]=y0; result[2]=y1; return result;};
+            qrem = y.divideAndRemainder(x); y = qrem[1];
+            y0 = y0.subtract(x0.multiply(qrem[0]));
+            y1 = y1.subtract(x1.multiply(qrem[0]));
+            if (y.equals(BigInteger.ZERO)) {result[0]=x; result[1]=x0; result[2]=x1; return result;};
+        }
+    }
+
+    public static void main(String[] args) {
```
Comment (kvaka313) [inline part2/Crypotology(Lab1)/src/main/java/lab1/ExtendedEuclidean.java]: exclude this method

File: part2/Crypotology(Lab1)/src/main/java/lab1/FastModPow.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
+package lab1;
```
Comment (kvaka313) [inline part2/Crypotology(Lab1)/src/main/java/lab1/FastModPow.java]: use Java Code Convention for naming packages

File: part2/Crypotology(Lab1)/src/main/java/lab1/Karatsuba.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.Random;
+
+// https://en.wikipedia.org/wiki/Karatsuba_algorithm
+public class Karatsuba {
+    private final static BigInteger ZERO = new BigInteger("0");
+
+    public static BigInteger karatsubaMultiply(BigInteger x, BigInteger y) {
+        // cutoff to brute force
+        int N = Math.max(x.bitLength(), y.bitLength());
+        if (N <= 25) return x.multiply(y);    // optimize this parameter
```
Comment (kvaka313) [inline part2/Crypotology(Lab1)/src/main/java/lab1/Karatsuba.java]: Use constant here

File: part2/Crypotology(Lab1)/src/main/java/lab1/MillerRabinTest.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    private static BigInteger uniformRandom(BigInteger bottom, BigInteger top) {
+        Random rnd = new Random();
+        BigInteger res;
+        do {
+            res = new BigInteger(top.bitLength(), rnd);
+        } while (res.compareTo(bottom) < 0 || res.compareTo(top) > 0);
+        return res;
+    }
+
+    public static void main(String[] args) {
```
Comment (kvaka313) [inline part2/Crypotology(Lab1)/src/main/java/lab1/MillerRabinTest.java]: exclude this method

File: part2/Crypotology(Lab1)/src/main/java/lab1/FermatTest.java (hunk 2 above)
Comment (Pilipets) [reply part2/Crypotology(Lab1)/src/main/java/lab1/FermatTest.java]: It's not a class for testing.
FermatTest class' functionality is tested in PrimesTest.java, but agree that naming is misleading. Will be changed.

File: part2/Crypotology(Lab1)/src/main/java/lab1/ExtendedEuclidean.java (hunk 4 above)
Comment (Pilipets) [reply part2/Crypotology(Lab1)/src/main/java/lab1/ExtendedEuclidean.java]: Done

File: part2/Crypotology(Lab1)/src/main/java/lab1/FastModPow.java (hunk 5 above)
Comment (Pilipets) [reply part2/Crypotology(Lab1)/src/main/java/lab1/FastModPow.java]: Done

File: part2/Crypotology(Lab1)/src/main/java/lab1/Karatsuba.java (hunk 6 above)
Comment (Pilipets) [reply part2/Crypotology(Lab1)/src/main/java/lab1/Karatsuba.java]: Agree, done.

File: part2/Crypotology(Lab1)/src/main/java/lab1/MillerRabinTest.java (hunk 7 above)
Comment (Pilipets) [reply part2/Crypotology(Lab1)/src/main/java/lab1/MillerRabinTest.java]: Done.

File: part2/Crypotology(Lab1)/src/main/java/lab1/Karatsuba.java (hunk 3 above)
Comment (Pilipets) [reply part2/Crypotology(Lab1)/src/main/java/lab1/Karatsuba.java]: Done.

File: part2/Crypotology(Lab1)/src/main/java/lab1/ExtendedEuclidean.java (hunk 1 above)
Comment (Pilipets) [reply part2/Crypotology(Lab1)/src/main/java/lab1/ExtendedEuclidean.java]: I agree about that.  Done.

**Pull request merged**